clang crashes when compiling aarch64 inline assembly with optimization enabled

## Environment

Ubuntu clang version 18.1.3 (1ubuntu1)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

operating system: ubuntu 24.04
clang is installed using apt.


## Minimal Reproduction Sample

https://godbolt.org/z/9YrKzxKrK
```c
// armtest.c
int main(void) {
  	register float a0 asm("v0");

  	for (int i = 0; i < 2; ++i) {
  		__asm__ volatile("nop" : [a0] "+w"(a0)::);
  	}

  	return 0;
}
```

compiling command:
```bash
clang --target=aarch64-linux-gnu armtest.c -O1
```

for comparison, gcc works well:
```bash
aarch64-linux-gnu-gcc armtest.c -O1
```


## Output
```
armtest.c:9:22: error: scalar-to-vector conversion failed, possible invalid constraint for vector type
    9 |                 __asm__ volatile("nop" : [a0] "+w"(a0)::);
      |                                  ^
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /usr/lib/llvm-18/bin/clang -cc1 -triple aarch64-unknown-linux-gnu -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name armtest.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=non-leaf -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu generic -target-feature +v8a -target-feature +fp-armv8 -target-feature +neon -target-abi aapcs -debugger-tuning=gdb -fdebug-compilation-dir=/home/shuhuai/test -fcoverage-compilation-dir=/home/shuhuai/test -resource-dir /usr/lib/llvm-18/lib/clang/18 -internal-isystem /usr/lib/llvm-18/lib/clang/18/include -internal-isystem /usr/local/include -internal-isystem /usr/bin/../lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -O1 -ferror-limit 19 -fno-signed-char -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -fcolor-diagnostics -target-feature +outline-atomics -target-feature -fmv -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/armtest-cd6e57.o -x c armtest.c
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'armtest.c'.
4.	Running pass 'Early Machine Loop Invariant Code Motion' on function '@main'
 #0 0x000078b8387a63bf llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xda63bf)
 #1 0x000078b8387a44f9 llvm::sys::RunSignalHandlers() (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xda44f9)
 #2 0x000078b8387a6b00 (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xda6b00)
 #3 0x000078b837245330 (/lib/x86_64-linux-gnu/libc.so.6+0x45330)
 #4 0x000078b83a686c08 (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0x2c86c08)
 #5 0x000078b838b9e9ce (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0x119e9ce)
 #6 0x000078b838b9d470 (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0x119d470)
 #7 0x000078b838b99400 (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0x1199400)
 #8 0x000078b838b826b9 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0x11826b9)
 #9 0x000078b8388f7772 llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xef7772)
#10 0x000078b8388fd2f4 llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xefd2f4)
#11 0x000078b8388f7e9f llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/lib/llvm-18/bin/../lib/libLLVM.so.18.1+0xef7e9f)
#12 0x000078b840e17310 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x1c17310)
#13 0x000078b84119fa07 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x1f9fa07)
#14 0x000078b83fd973d6 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0xb973d6)
#15 0x000078b841c0662c clang::FrontendAction::Execute() (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x2a0662c)
#16 0x000078b841b830b4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x29830b4)
#17 0x000078b841c8263d clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18.1+0x2a8263d)
#18 0x0000573e0e1fa42e cc1_main(llvm::ArrayRef<char const*>, char const*, void*) build-llvm/tools/clang/stage2-bins/clang/tools/driver/cc1_main.cpp:294:15
#19 0x0000573e0e1f7894 ExecuteCC1Tool build-llvm/tools/clang/stage2-bins/clang/tools/driver/driver.cpp:365:12
#20 0x0000573e0e1f6a26 clang_main(int, char**, llvm::ToolContext const&) build-llvm/tools/clang/stage2-bins/clang/tools/driver/driver.cpp:405:12
#21 0x0000573e0e204383 main build-llvm/tools/clang/stage2-bins/build-llvm/tools/clang/stage2-bins/tools/clang/tools/driver/clang-driver.cpp:17:10
#22 0x000078b83722a1ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#23 0x000078b83722a28b call_init ./csu/../csu/libc-start.c:128:20
#24 0x000078b83722a28b __libc_start_main ./csu/../csu/libc-start.c:347:5
#25 0x0000573e0e1f4255 _start (/usr/lib/llvm-18/bin/clang+0xe255)
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Ubuntu clang version 18.1.3 (1ubuntu1)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/armtest-3b582d.c
clang: note: diagnostic msg: /tmp/armtest-3b582d.sh
clang: note: diagnostic msg: 

********************
```

## Additional Notes

1. set optimization level to `-O0` will not cause crash
2. another way to avoid crash is to include `arm_neon.h` and change the type of `a0` from `float` to `float32x4_t`
3. I think the error message may be misleading since the cursor in the message is pointed to the `nop` instruction. It crashes no matter what the instructions are.

[diagnostic_msg.zip](https://github.com/user-attachments/files/21758221/diagnostic_msg.zip)

--- a/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
+++ b/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
@@ -13362,12 +13362,28 @@ AArch64TargetLowering::getRegForInlineAsmConstraint(
       int RegNo;
       bool Failed = Constraint.slice(2, Size - 1).getAsInteger(10, RegNo);
       if (!Failed && RegNo >= 0 && RegNo <= 31) {
-        // v0 - v31 are aliases of q0 - q31 or d0 - d31 depending on size.
+        // v0 - v31 are aliases of q0/d0/s0/h0 - ...31 depending on size.
         // By default we'll emit v0-v31 for this unless there's a modifier where
         // we'll emit the correct register as well.
-        if (VT != MVT::Other && VT.getSizeInBits() <= 64) {
-          Res.first = AArch64::FPR64RegClass.getRegister(RegNo);
-          Res.second = &AArch64::FPR64RegClass;
+        if (VT != MVT::Other) {
+          switch (VT.getSizeInBits()) {
+          default:
+            Res.first = AArch64::FPR128RegClass.getRegister(RegNo);
+            Res.second = &AArch64::FPR128RegClass;
+            break;
+          case 64:
+            Res.first = AArch64::FPR64RegClass.getRegister(RegNo);
+            Res.second = &AArch64::FPR64RegClass;
+            break;
+          case 32:
+            Res.first = AArch64::FPR32RegClass.getRegister(RegNo);
+            Res.second = &AArch64::FPR32RegClass;
+            break;
+          case 16:
+            Res.first = AArch64::FPR16RegClass.getRegister(RegNo);
+            Res.second = &AArch64::FPR16RegClass;
+            break;
+          }
         } else {
           Res.first = AArch64::FPR128RegClass.getRegister(RegNo);
           Res.second = &AArch64::FPR128RegClass;

--- a/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
+++ b/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
@@ -13367,22 +13367,24 @@ AArch64TargetLowering::getRegForInlineAsmConstraint(
         // we'll emit the correct register as well.
         if (VT != MVT::Other) {
           switch (VT.getSizeInBits()) {
-          default:
-            Res.first = AArch64::FPR128RegClass.getRegister(RegNo);
-            Res.second = &AArch64::FPR128RegClass;
-            break;
-          case 64:
-            Res.first = AArch64::FPR64RegClass.getRegister(RegNo);
-            Res.second = &AArch64::FPR64RegClass;
+          case 16:
+            Res.first = AArch64::FPR16RegClass.getRegister(RegNo);
+            Res.second = &AArch64::FPR16RegClass;
             break;
           case 32:
             Res.first = AArch64::FPR32RegClass.getRegister(RegNo);
             Res.second = &AArch64::FPR32RegClass;
             break;
-          case 16:
-            Res.first = AArch64::FPR16RegClass.getRegister(RegNo);
-            Res.second = &AArch64::FPR16RegClass;
+          case 64:
+            Res.first = AArch64::FPR64RegClass.getRegister(RegNo);
+            Res.second = &AArch64::FPR64RegClass;
             break;
+          case 128:
+            Res.first = AArch64::FPR128RegClass.getRegister(RegNo);
+            Res.second = &AArch64::FPR128RegClass;
+            break;
+          default:
+            return std::make_pair(0U, nullptr);
           }
         } else {
           Res.first = AArch64::FPR128RegClass.getRegister(RegNo);

--- a/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
+++ b/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
@@ -13365,7 +13365,7 @@ AArch64TargetLowering::getRegForInlineAsmConstraint(
         // v0 - v31 are aliases of q0 - q31 or d0 - d31 depending on size.
         // By default we'll emit v0-v31 for this unless there's a modifier where
         // we'll emit the correct register as well.
-        if (VT != MVT::Other && VT.getSizeInBits() == 64) {
+        if (VT != MVT::Other && VT.getSizeInBits() <= 64) {
           Res.first = AArch64::FPR64RegClass.getRegister(RegNo);
           Res.second = &AArch64::FPR64RegClass;
         } else {

--- a/llvm/test/CodeGen/AArch64/inline-asm-nop-reg.ll
+++ b/llvm/test/CodeGen/AArch64/inline-asm-nop-reg.ll
@@ -1,15 +1,16 @@
 ; RUN: llc -O1 -mtriple=aarch64-linux-gnu %s -o - 2>&1 | FileCheck %s
 
-; This test checks that the code containing "nop" inline assembler instruction
-; with 16/32/64-bit FP "v0" register, will be successfully compiled
-; and generated code will contain one optimized nop-instruction
-; per each function.
+; This test contains "nop" inline assembler instruction with 16..128-bit
+; affectd FP "v0" register. It checks that the IR will be successfully
+; compiled and generated code will contain nop-instruction and proper register
+; name placed in each function.
 ;
 ; IR for this test was generated from the following source code:
 ;
 ; #define _FP16 _Float16
 ; #define _FP32 float
 ; #define _FP64 double
+; #define _FP128 long double
 ;
 ; #define FOO(BITS) \
 ; int foo##BITS(void) { \
@@ -23,98 +24,134 @@
 ; FOO(16)
 ; FOO(32)
 ; FOO(64)
+; FOO(128)
 
 
 ; test nop_fp16_reg
 ; CHECK-LABEL: foo16:
 ; CHECK: nop
-; CHECK-NOT: nop
+; CHECK: str h0
 define dso_local i32 @foo16() #0 {
-  %1 = alloca half, align 2
-  %2 = alloca i32, align 4
-  store i32 0, ptr %2, align 4
-  br label %3
+entry:
+  %a0 = alloca half, align 2
+  %i = alloca i32, align 4
+  store i32 0, ptr %i, align 4
+  br label %for.cond
 
-3:                                                ; preds = %9, %0
-  %4 = load i32, ptr %2, align 4
-  %5 = icmp slt i32 %4, 2
-  br i1 %5, label %6, label %12
+for.cond:                                         ; preds = %for.inc, %entry
+  %0 = load i32, ptr %i, align 4
+  %cmp = icmp slt i32 %0, 2
+  br i1 %cmp, label %for.body, label %for.end
 
-6:                                                ; preds = %3
-  %7 = load half, ptr %1, align 2
-  %8 = call half asm sideeffect "nop", "={v0},{v0}"(half %7) #1, !srcloc !6
-  store half %8, ptr %1, align 2
-  br label %9
+for.body:                                         ; preds = %for.cond
+  %1 = load half, ptr %a0, align 2
+  %2 = call half asm sideeffect "nop", "={v0},{v0}"(half %1) #1, !srcloc !6
+  store half %2, ptr %a0, align 2
+  br label %for.inc
 
-9:                                                ; preds = %6
-  %10 = load i32, ptr %2, align 4
-  %11 = add nsw i32 %10, 1
-  store i32 %11, ptr %2, align 4
-  br label %3, !llvm.loop !7
+for.inc:                                          ; preds = %for.body
+  %3 = load i32, ptr %i, align 4
+  %inc = add nsw i32 %3, 1
+  store i32 %inc, ptr %i, align 4
+  br label %for.cond, !llvm.loop !7
 
-12:                                               ; preds = %3
+for.end:                                          ; preds = %for.cond
   ret i32 0
 }
 
 ; test nop_fp32_reg
 ; CHECK-LABEL: foo32:
 ; CHECK: nop
-; CHECK-NOT: nop
+; CHECK: str s0
 define dso_local i32 @foo32() #0 {
-  %1 = alloca float, align 4
-  %2 = alloca i32, align 4
-  store i32 0, ptr %2, align 4
-  br label %3
+entry:
+  %a0 = alloca float, align 4
+  %i = alloca i32, align 4
+  store i32 0, ptr %i, align 4
+  br label %for.cond
 
-3:                                                ; preds = %9, %0
-  %4 = load i32, ptr %2, align 4
-  %5 = icmp slt i32 %4, 2
-  br i1 %5, label %6, label %12
+for.cond:                                         ; preds = %for.inc, %entry
+  %0 = load i32, ptr %i, align 4
+  %cmp = icmp slt i32 %0, 2
+  br i1 %cmp, label %for.body, label %for.end
 
-6:                                                ; preds = %3
-  %7 = load float, ptr %1, align 4
-  %8 = call float asm sideeffect "nop", "={v0},{v0}"(float %7) #1, !srcloc !9
-  store float %8, ptr %1, align 4
-  br label %9
+for.body:                                         ; preds = %for.cond
+  %1 = load float, ptr %a0, align 4
+  %2 = call float asm sideeffect "nop", "={v0},{v0}"(float %1) #1, !srcloc !9
+  store float %2, ptr %a0, align 4
+  br label %for.inc
 
-9:                                                ; preds = %6
-  %10 = load i32, ptr %2, align 4
-  %11 = add nsw i32 %10, 1
-  store i32 %11, ptr %2, align 4
-  br label %3, !llvm.loop !10
+for.inc:                                          ; preds = %for.body
+  %3 = load i32, ptr %i, align 4
+  %inc = add nsw i32 %3, 1
+  store i32 %inc, ptr %i, align 4
+  br label %for.cond, !llvm.loop !10
 
-12:                                               ; preds = %3
+for.end:                                          ; preds = %for.cond
   ret i32 0
 }
 
 ; test nop_fp64_reg
 ; CHECK-LABEL: foo64:
 ; CHECK: nop
-; CHECK-NOT: nop
+; CHECK: str d0
 define dso_local i32 @foo64() #0 {
-  %1 = alloca double, align 8
-  %2 = alloca i32, align 4
-  store i32 0, ptr %2, align 4
-  br label %3
+entry:
+  %a0 = alloca double, align 8
+  %i = alloca i32, align 4
+  store i32 0, ptr %i, align 4
+  br label %for.cond
 
-3:                                                ; preds = %9, %0
-  %4 = load i32, ptr %2, align 4
-  %5 = icmp slt i32 %4, 2
-  br i1 %5, label %6, label %12
+for.cond:                                         ; preds = %for.inc, %entry
+  %0 = load i32, ptr %i, align 4
+  %cmp = icmp slt i32 %0, 2
+  br i1 %cmp, label %for.body, label %for.end
 
-6:                                                ; preds = %3
-  %7 = load double, ptr %1, align 8
-  %8 = call double asm sideeffect "nop", "={v0},{v0}"(double %7) #1, !srcloc !11
-  store double %8, ptr %1, align 8
-  br label %9
+for.body:                                         ; preds = %for.cond
+  %1 = load double, ptr %a0, align 8
+  %2 = call double asm sideeffect "nop", "={v0},{v0}"(double %1) #1, !srcloc !11
+  store double %2, ptr %a0, align 8
+  br label %for.inc
 
-9:                                                ; preds = %6
-  %10 = load i32, ptr %2, align 4
-  %11 = add nsw i32 %10, 1
-  store i32 %11, ptr %2, align 4
-  br label %3, !llvm.loop !12
+for.inc:                                          ; preds = %for.body
+  %3 = load i32, ptr %i, align 4
+  %inc = add nsw i32 %3, 1
+  store i32 %inc, ptr %i, align 4
+  br label %for.cond, !llvm.loop !12
 
-12:                                               ; preds = %3
+for.end:                                          ; preds = %for.cond
+  ret i32 0
+}
+
+; test nop_fp128_reg
+; CHECK-LABEL: foo128:
+; CHECK: nop
+; CHECK: str q0
+define dso_local i32 @foo128() #0 {
+entry:
+  %a0 = alloca fp128, align 16
+  %i = alloca i32, align 4
+  store i32 0, ptr %i, align 4
+  br label %for.cond
+
+for.cond:                                         ; preds = %for.inc, %entry
+  %0 = load i32, ptr %i, align 4
+  %cmp = icmp slt i32 %0, 2
+  br i1 %cmp, label %for.body, label %for.end
+
+for.body:                                         ; preds = %for.cond
+  %1 = load fp128, ptr %a0, align 16
+  %2 = call fp128 asm sideeffect "nop", "={v0},{v0}"(fp128 %1) #1, !srcloc !13
+  store fp128 %2, ptr %a0, align 16
+  br label %for.inc
+
+for.inc:                                          ; preds = %for.body
+  %3 = load i32, ptr %i, align 4
+  %inc = add nsw i32 %3, 1
+  store i32 %inc, ptr %i, align 4
+  br label %for.cond, !llvm.loop !14
+
+for.end:                                          ; preds = %for.cond
   ret i32 0
 }
 
@@ -125,12 +162,14 @@ define dso_local i32 @foo64() #0 {
 !1 = !{i32 8, !"PIC Level", i32 2}
 !2 = !{i32 7, !"PIE Level", i32 2}
 !3 = !{i32 7, !"uwtable", i32 2}
-!4 = !{i32 7, !"frame-pointer", i32 1}
-!5 = !{!"clang version 22.0.0git"}
-!6 = !{i64 2147502427}
+!4 = !{i32 7, !"frame-pointer", i32 4}
+!5 = !{!"clang version 22.0.0git (https://github.com/llvm/llvm-project.git)"}
+!6 = !{i64 2147502487}
 !7 = distinct !{!7, !8}
 !8 = !{!"llvm.loop.mustprogress"}
-!9 = !{i64 2147502622}
+!9 = !{i64 2147502682}
 !10 = distinct !{!10, !8}
-!11 = !{i64 2147502814}
+!11 = !{i64 2147502874}
 !12 = distinct !{!12, !8}
+!13 = !{i64 2147503067}
+!14 = distinct !{!14, !8}

--- a/llvm/test/CodeGen/AArch64/inline-asm-nop-reg.ll
+++ b/llvm/test/CodeGen/AArch64/inline-asm-nop-reg.ll
@@ -1,158 +1,150 @@
 ; RUN: llc -O1 -mtriple=aarch64-linux-gnu %s -o - 2>&1 | FileCheck %s
 
 ; This test contains "nop" inline assembler instruction with 16..128-bit
-; affectd FP "v0" register. It checks that the IR will be successfully
+; affectd "v0" register. It checks that the IR will be successfully
 ; compiled and generated code will contain nop-instruction and proper register
 ; name placed in each function.
 ;
 ; IR for this test was generated from the following source code:
+;
+; #include <stdint.h>
 ;
 ; #define _FP16 _Float16
 ; #define _FP32 float
 ; #define _FP64 double
 ; #define _FP128 long double
 ;
-; #define FOO(BITS) \
-; int foo##BITS(void) { \
-;   register _FP##BITS a0 asm("v0"); \
-;   for (int i = 0; i < 2; ++i) { \
-;     __asm__ volatile("nop" : [a0] "+w"(a0)::); \
-;   } \
-;   return 0; \
+; #define _INT16 int16_t
+; #define _INT32 int32_t
+; #define _INT64 int64_t
+; #define _INT128 __int128
+;
+; #define FOO(TYPE) \
+; void foo_##TYPE(void) { \
+;   register _##TYPE a0 asm("v0"); \
+;   __asm__ volatile("nop" : [a0] "+w"(a0)::); \
 ; }
 ;
-; FOO(16)
-; FOO(32)
-; FOO(64)
-; FOO(128)
+; FOO(FP16)
+; FOO(FP32)
+; FOO(FP64)
+; FOO(FP128)
+;
+; FOO(INT16)
+; FOO(INT32)
+; FOO(INT64)
+; FOO(INT128)
 
-
-; test nop_fp16_reg
-; CHECK-LABEL: foo16:
+; test nop_FP16_reg
+; CHECK-LABEL: foo_FP16:
+; CHECK: ldr h0
 ; CHECK: nop
 ; CHECK: str h0
-define dso_local i32 @foo16() #0 {
+define dso_local void @foo_FP16() #0 {
 entry:
   %a0 = alloca half, align 2
-  %i = alloca i32, align 4
-  store i32 0, ptr %i, align 4
-  br label %for.cond
-
-for.cond:                                         ; preds = %for.inc, %entry
-  %0 = load i32, ptr %i, align 4
-  %cmp = icmp slt i32 %0, 2
-  br i1 %cmp, label %for.body, label %for.end
-
-for.body:                                         ; preds = %for.cond
-  %1 = load half, ptr %a0, align 2
-  %2 = call half asm sideeffect "nop", "={v0},{v0}"(half %1) #1, !srcloc !6
-  store half %2, ptr %a0, align 2
-  br label %for.inc
-
-for.inc:                                          ; preds = %for.body
-  %3 = load i32, ptr %i, align 4
-  %inc = add nsw i32 %3, 1
-  store i32 %inc, ptr %i, align 4
-  br label %for.cond, !llvm.loop !7
-
-for.end:                                          ; preds = %for.cond
-  ret i32 0
+  %0 = load half, ptr %a0, align 2
+  %1 = call half asm sideeffect "nop", "={v0},{v0}"(half %0) #1, !srcloc !6
+  store half %1, ptr %a0, align 2
+  ret void
 }
 
-; test nop_fp32_reg
-; CHECK-LABEL: foo32:
+; test nop_FP32_reg
+; CHECK-LABEL: foo_FP32:
+; CHECK: ldr s0
 ; CHECK: nop
 ; CHECK: str s0
-define dso_local i32 @foo32() #0 {
+define dso_local void @foo_FP32() #0 {
 entry:
   %a0 = alloca float, align 4
-  %i = alloca i32, align 4
-  store i32 0, ptr %i, align 4
-  br label %for.cond
-
-for.cond:                                         ; preds = %for.inc, %entry
-  %0 = load i32, ptr %i, align 4
-  %cmp = icmp slt i32 %0, 2
-  br i1 %cmp, label %for.body, label %for.end
-
-for.body:                                         ; preds = %for.cond
-  %1 = load float, ptr %a0, align 4
-  %2 = call float asm sideeffect "nop", "={v0},{v0}"(float %1) #1, !srcloc !9
-  store float %2, ptr %a0, align 4
-  br label %for.inc
-
-for.inc:                                          ; preds = %for.body
-  %3 = load i32, ptr %i, align 4
-  %inc = add nsw i32 %3, 1
-  store i32 %inc, ptr %i, align 4
-  br label %for.cond, !llvm.loop !10
-
-for.end:                                          ; preds = %for.cond
-  ret i32 0
+  %0 = load float, ptr %a0, align 4
+  %1 = call float asm sideeffect "nop", "={v0},{v0}"(float %0) #1, !srcloc !7
+  store float %1, ptr %a0, align 4
+  ret void
 }
 
-; test nop_fp64_reg
-; CHECK-LABEL: foo64:
+; test nop_FP64_reg
+; CHECK-LABEL: foo_FP64:
+; CHECK: ldr d0
 ; CHECK: nop
 ; CHECK: str d0
-define dso_local i32 @foo64() #0 {
+define dso_local void @foo_FP64() #0 {
 entry:
   %a0 = alloca double, align 8
-  %i = alloca i32, align 4
-  store i32 0, ptr %i, align 4
-  br label %for.cond
-
-for.cond:                                         ; preds = %for.inc, %entry
-  %0 = load i32, ptr %i, align 4
-  %cmp = icmp slt i32 %0, 2
-  br i1 %cmp, label %for.body, label %for.end
-
-for.body:                                         ; preds = %for.cond
-  %1 = load double, ptr %a0, align 8
-  %2 = call double asm sideeffect "nop", "={v0},{v0}"(double %1) #1, !srcloc !11
-  store double %2, ptr %a0, align 8
-  br label %for.inc
-
-for.inc:                                          ; preds = %for.body
-  %3 = load i32, ptr %i, align 4
-  %inc = add nsw i32 %3, 1
-  store i32 %inc, ptr %i, align 4
-  br label %for.cond, !llvm.loop !12
-
-for.end:                                          ; preds = %for.cond
-  ret i32 0
+  %0 = load double, ptr %a0, align 8
+  %1 = call double asm sideeffect "nop", "={v0},{v0}"(double %0) #1, !srcloc !8
+  store double %1, ptr %a0, align 8
+  ret void
 }
 
-; test nop_fp128_reg
-; CHECK-LABEL: foo128:
+; test nop_FP128_reg
+; CHECK-LABEL: foo_FP128:
+; CHECK: ldr q0
 ; CHECK: nop
 ; CHECK: str q0
-define dso_local i32 @foo128() #0 {
+define dso_local void @foo_FP128() #0 {
 entry:
   %a0 = alloca fp128, align 16
-  %i = alloca i32, align 4
-  store i32 0, ptr %i, align 4
-  br label %for.cond
+  %0 = load fp128, ptr %a0, align 16
+  %1 = call fp128 asm sideeffect "nop", "={v0},{v0}"(fp128 %0) #1, !srcloc !9
+  store fp128 %1, ptr %a0, align 16
+  ret void
+}
 
-for.cond:                                         ; preds = %for.inc, %entry
-  %0 = load i32, ptr %i, align 4
-  %cmp = icmp slt i32 %0, 2
-  br i1 %cmp, label %for.body, label %for.end
+; test nop_INT16_reg
+; CHECK-LABEL: foo_INT16:
+; CHECK: ldr h0
+; CHECK: nop
+; CHECK: str h0
+define dso_local void @foo_INT16() #0 {
+entry:
+  %a0 = alloca i16, align 2
+  %0 = load i16, ptr %a0, align 2
+  %1 = call i16 asm sideeffect "nop", "={v0},{v0}"(i16 %0) #1, !srcloc !10
+  store i16 %1, ptr %a0, align 2
+  ret void
+}
 
-for.body:                                         ; preds = %for.cond
-  %1 = load fp128, ptr %a0, align 16
-  %2 = call fp128 asm sideeffect "nop", "={v0},{v0}"(fp128 %1) #1, !srcloc !13
-  store fp128 %2, ptr %a0, align 16
-  br label %for.inc
+; test nop_INT32_reg
+; CHECK-LABEL: foo_INT32:
+; CHECK: ldr s0
+; CHECK: nop
+; CHECK: str s0
+define dso_local void @foo_INT32() #0 {
+entry:
+  %a0 = alloca i32, align 4
+  %0 = load i32, ptr %a0, align 4
+  %1 = call i32 asm sideeffect "nop", "={v0},{v0}"(i32 %0) #1, !srcloc !11
+  store i32 %1, ptr %a0, align 4
+  ret void
+}
 
-for.inc:                                          ; preds = %for.body
-  %3 = load i32, ptr %i, align 4
-  %inc = add nsw i32 %3, 1
-  store i32 %inc, ptr %i, align 4
-  br label %for.cond, !llvm.loop !14
+; test nop_INT64_reg
+; CHECK-LABEL: foo_INT64:
+; CHECK: ldr d0
+; CHECK: nop
+; CHECK: str d0
+define dso_local void @foo_INT64() #0 {
+entry:
+  %a0 = alloca i64, align 8
+  %0 = load i64, ptr %a0, align 8
+  %1 = call i64 asm sideeffect "nop", "={v0},{v0}"(i64 %0) #1, !srcloc !12
+  store i64 %1, ptr %a0, align 8
+  ret void
+}
 
-for.end:                                          ; preds = %for.cond
-  ret i32 0
+; test nop_INT128_reg
+; CHECK-LABEL: foo_INT128:
+; CHECK: ldr q0
+; CHECK: nop
+; CHECK: str q0
+define dso_local void @foo_INT128() #0 {
+entry:
+  %a0 = alloca i128, align 16
+  %0 = load i128, ptr %a0, align 16
+  %1 = call i128 asm sideeffect "nop", "={v0},{v0}"(i128 %0) #1, !srcloc !13
+  store i128 %1, ptr %a0, align 16
+  ret void
 }
 
 !llvm.module.flags = !{!0, !1, !2, !3, !4}
@@ -164,12 +156,11 @@ for.end:                                          ; preds = %for.cond
 !3 = !{i32 7, !"uwtable", i32 2}
 !4 = !{i32 7, !"frame-pointer", i32 4}
 !5 = !{!"clang version 22.0.0git (https://github.com/llvm/llvm-project.git)"}
-!6 = !{i64 2147502487}
-!7 = distinct !{!7, !8}
-!8 = !{!"llvm.loop.mustprogress"}
-!9 = !{i64 2147502682}
-!10 = distinct !{!10, !8}
-!11 = !{i64 2147502874}
-!12 = distinct !{!12, !8}
-!13 = !{i64 2147503067}
-!14 = distinct !{!14, !8}
+!6 = !{i64 2147598584}
+!7 = !{i64 2147598730}
+!8 = !{i64 2147598873}
+!9 = !{i64 2147599017}
+!10 = !{i64 2147599170}
+!11 = !{i64 2147599319}
+!12 = !{i64 2147599468}
+!13 = !{i64 2147599617}

--- a/llvm/test/CodeGen/AArch64/inline-asm-nop-reg.ll
+++ b/llvm/test/CodeGen/AArch64/inline-asm-nop-reg.ll
@@ -1,0 +1,136 @@
+; RUN: llc -O1 -mtriple=aarch64-linux-gnu %s -o - 2>&1 | FileCheck %s
+
+; This test checks that the code containing "nop" inline assembler instruction
+; with 16/32/64-bit FP "v0" register, will be successfully compiled
+; and generated code will contain one optimized nop-instruction
+; per each function.
+;
+; IR for this test was generated from the following source code:
+;
+; #define _FP16 _Float16
+; #define _FP32 float
+; #define _FP64 double
+;
+; #define FOO(BITS) \
+; int foo##BITS(void) { \
+;   register _FP##BITS a0 asm("v0"); \
+;   for (int i = 0; i < 2; ++i) { \
+;     __asm__ volatile("nop" : [a0] "+w"(a0)::); \
+;   } \
+;   return 0; \
+; }
+;
+; FOO(16)
+; FOO(32)
+; FOO(64)
+
+
+; test nop_fp16_reg
+; CHECK-LABEL: foo16:
+; CHECK: nop
+; CHECK-NOT: nop
+define dso_local i32 @foo16() #0 {
+  %1 = alloca half, align 2
+  %2 = alloca i32, align 4
+  store i32 0, ptr %2, align 4
+  br label %3
+
+3:                                                ; preds = %9, %0
+  %4 = load i32, ptr %2, align 4
+  %5 = icmp slt i32 %4, 2
+  br i1 %5, label %6, label %12
+
+6:                                                ; preds = %3
+  %7 = load half, ptr %1, align 2
+  %8 = call half asm sideeffect "nop", "={v0},{v0}"(half %7) #1, !srcloc !6
+  store half %8, ptr %1, align 2
+  br label %9
+
+9:                                                ; preds = %6
+  %10 = load i32, ptr %2, align 4
+  %11 = add nsw i32 %10, 1
+  store i32 %11, ptr %2, align 4
+  br label %3, !llvm.loop !7
+
+12:                                               ; preds = %3
+  ret i32 0
+}
+
+; test nop_fp32_reg
+; CHECK-LABEL: foo32:
+; CHECK: nop
+; CHECK-NOT: nop
+define dso_local i32 @foo32() #0 {
+  %1 = alloca float, align 4
+  %2 = alloca i32, align 4
+  store i32 0, ptr %2, align 4
+  br label %3
+
+3:                                                ; preds = %9, %0
+  %4 = load i32, ptr %2, align 4
+  %5 = icmp slt i32 %4, 2
+  br i1 %5, label %6, label %12
+
+6:                                                ; preds = %3
+  %7 = load float, ptr %1, align 4
+  %8 = call float asm sideeffect "nop", "={v0},{v0}"(float %7) #1, !srcloc !9
+  store float %8, ptr %1, align 4
+  br label %9
+
+9:                                                ; preds = %6
+  %10 = load i32, ptr %2, align 4
+  %11 = add nsw i32 %10, 1
+  store i32 %11, ptr %2, align 4
+  br label %3, !llvm.loop !10
+
+12:                                               ; preds = %3
+  ret i32 0
+}
+
+; test nop_fp64_reg
+; CHECK-LABEL: foo64:
+; CHECK: nop
+; CHECK-NOT: nop
+define dso_local i32 @foo64() #0 {
+  %1 = alloca double, align 8
+  %2 = alloca i32, align 4
+  store i32 0, ptr %2, align 4
+  br label %3
+
+3:                                                ; preds = %9, %0
+  %4 = load i32, ptr %2, align 4
+  %5 = icmp slt i32 %4, 2
+  br i1 %5, label %6, label %12
+
+6:                                                ; preds = %3
+  %7 = load double, ptr %1, align 8
+  %8 = call double asm sideeffect "nop", "={v0},{v0}"(double %7) #1, !srcloc !11
+  store double %8, ptr %1, align 8
+  br label %9
+
+9:                                                ; preds = %6
+  %10 = load i32, ptr %2, align 4
+  %11 = add nsw i32 %10, 1
+  store i32 %11, ptr %2, align 4
+  br label %3, !llvm.loop !12
+
+12:                                               ; preds = %3
+  ret i32 0
+}
+
+!llvm.module.flags = !{!0, !1, !2, !3, !4}
+!llvm.ident = !{!5}
+
+!0 = !{i32 1, !"wchar_size", i32 4}
+!1 = !{i32 8, !"PIC Level", i32 2}
+!2 = !{i32 7, !"PIE Level", i32 2}
+!3 = !{i32 7, !"uwtable", i32 2}
+!4 = !{i32 7, !"frame-pointer", i32 1}
+!5 = !{!"clang version 22.0.0git"}
+!6 = !{i64 2147502427}
+!7 = distinct !{!7, !8}
+!8 = !{!"llvm.loop.mustprogress"}
+!9 = !{i64 2147502622}
+!10 = distinct !{!10, !8}
+!11 = !{i64 2147502814}
+!12 = distinct !{!12, !8}

